### ISSUE TITLE
Image in Helm Job template isn't replaced with correct tagged image

### ISSUE THREAD

**Issue by charlesg99** [GoogleContainerTools/skaffold#8303]
Just migrated from Skaffold 1 to 2. Using Helm to deploy and it seems like the templates containing a `kind: Job` don't have the correct skaffold image injected. The `kind: Deployment` in the same project work as expected. This used to work fine with version 1 and artifactOverrides.

Here are some example yamls

skaffold.yaml
```
apiVersion: skaffold/v4beta1
kind: Config
build:
  artifacts:
  - image: backend-image
    context: ../src/backend
  - image: job-image
    context: ../src/job
  tagPolicy:
    gitCommit:
      variant: CommitSha
manifests:
  helm:
    flags:
      upgrade:
      - --cleanup-on-fail
    releases:
    - name: project
      chartPath: helm
      valuesFiles:
      - helm/values.yaml
      - helm/values.yaml.override
      setValues:
        backendImage: backend-image          <---------------------------- this one works fine
        jobImage: job-image
deploy:
  helm: {}
```

Job template:
```
apiVersion: batch/v1
kind: Job
metadata:
  name: {{ .Release.Name }}-job
  labels:
spec:
  template:
    metadata:
    spec:
      containers:
      - name: {{ .Release.Name }}-image-job
        image: {{ .Values.jobImage }}
```

In the console output I can see that the "render" doesn't set the image with the tag (`job-image:XXXXXXXXXXXXXXXXXX`), it stays to `job-image` in the assembled yaml. In the same example the render output shows that backend-image:XXXXXXXXXXXXXX is present in a regular deployment template.

Also the output seems to load the correct tagged image (locally using Kind, but same thing happens in a live cluster):

```
Tags used in deployment:
 - backend-image -> backend-image:89418bfa46eec750ec3443b56f7d4dc81058f4f3f2506c8cc4db62810ba34643
 - job-image -> job-image:8c73ab625f7434b92db0eac4a5b4a19f928bfc24d0ba23992e753994e8aa925c
```
 
```
 - job-image:8c73ab625f7434b92db0eac4a5b4a19f928bfc24d0ba23992e753994e8aa925c -> DEBU[0093] Running command: [kind load docker-image --name project job-image:8c73ab625f7434b92db0eac4a5b4a19f928bfc24d0ba23992e753994e8aa925c]  subtask=0 task=Deploy
DEBU[0101] Command output: [], stderr: Image: "job-image:8c73ab625f7434b92db0eac4a5b4a19f928bfc24d0ba23992e753994e8aa925c" with ID "sha256:8c73ab625f7434b92db0eac4a5b4a19f928bfc24d0ba23992e753994e8aa925c" not yet present on node "project-control-plane", loading...
Image: "job-image:8c73ab625f7434b92db0eac4a5b4a19f928bfc24d0ba23992e753994e8aa925c" with ID "sha256:8c73ab625f7434b92db0eac4a5b4a19f928bfc24d0ba23992e753994e8aa925c" not yet present on node "project-worker", loading...  subtask=0 task=Deploy
```

**Comment by aaron-prindle** [comment]
@charlesg99 thanks for flagging this.  Can you try using the recommended approach of using `setValueTemplates` (vs `setValues`):
https://skaffold.dev/docs/pipeline-stages/deployers/helm/

change:
```
manifests:
  helm:
    flags:
      upgrade:
      - --cleanup-on-fail
    releases:
    - name: project
      chartPath: helm
      valuesFiles:
      - helm/values.yaml
      - helm/values.yaml.override
      setValues:
        backendImage: backend-image
        jobImage: job-image
```
to:

```
manifests:
  helm:
    flags:
      upgrade:
      - --cleanup-on-fail
    releases:
    - name: project
      chartPath: helm
      valuesFiles:
      - helm/values.yaml
      - helm/values.yaml.override
      setValues:
        backendImage: "{{.IMAGE_FULLY_QUALIFIED_backend_image}}"
        jobImage: "{{.IMAGE_FULLY_QUALIFIED_job_image}}"
```

Also what version of Skaffold were you using here eg( `v2.0.?`) and did you generate this skaffold.yalm using `skaffold fix`?

**Comment by aaron-prindle** [comment]
In my testing using `setValueTemplates` I am able to get this to work properly IIUC:
```
apiVersion: skaffold/v4beta1
kind: Config
build:
  artifacts:
  - image: skaffold-helm
    context: skaffold-helm
  - image: skaffold-helm-2
    context: skaffold-helm-2
deploy:
  helm:
    releases:
    - name: skaffold-helm
      chartPath: charts
      setValueTemplates:
        image: "{{.IMAGE_FULLY_QUALIFIED_skaffold_helm}}"
        image2: "{{.IMAGE_FULLY_QUALIFIED_skaffold_helm_2}}"
```

```
apiVersion: batch/v1
kind: Job
metadata:
  name: {{ .Chart.Name }}
spec:
  template:
    metadata:
      labels: {}
    spec:
      containers:
      - name: {{ .Chart.Name }}
        image: {{ .Values.image }}
      restartPolicy: Never
  backoffLimit: 4
---
apiVersion: batch/v1
kind: Job
metadata:
  name: {{ .Chart.Name }}-2
spec:
  template:
    metadata:
      labels: {}
    spec:
      containers:
      - name: {{ .Chart.Name }}
        image: {{ .Values.image2 }}
      restartPolicy: Never
  backoffLimit: 4
```

gist with full output of `skaffold dev --default-repo=<my-repo>`
https://gist.github.com/aaron-prindle/ccc9190ca2dcf9f7bc67f2e22a1c4372

**Comment by charlesg99** [comment]
Thanks Aaron, it works with setValueTemplates. I did use `skaffold fix` to migrate from version 1 to 2 but apparently it was before the introduction of setValueTemplates? I remember reading about the migration process in another issue and setValueTemplates wasn't yet mentionned. In my CI I'm installing using the latest stable binary url. I guess this can be closed.

**Comment by aaron-prindle** [comment]
Thanks for the info @charlesg99, the recommended method changed from `v2.0.2` -> `v2.0.4` so this is likely the cause of confusion.  Sorry for the inconvenience, I am planning to make this more robust with some validations, etc. but the goal for now is that `skaffold fix` works seamlessly for `v1` -> `v2.0.4+` w/ the caveat that if you used `skaffold fix` prior to `v2.0.4` (`v2.0.3`) there might be some issues like the one you encountered

**Comment by ivanovaleksey** [comment]
hey @aaron-prindle , by the way, is it by design that image injection works automatically for `kind: Deployment` and doesn't work for `kind: Job`?
